### ISSUE TITLE
Add an option for taint analysis to analyze taint of global variables across files?

### ISSUE THREAD

**Issue by TysonAndre** [vimeo/psalm#3778]
Expected: There should be a way to warn about this in the latest commit of psalm
Observed: `psalm --taint-analysis --threads=1` does not warn.

https://psalm.dev/docs/running_psalm/configuration/ mentions no settings related to taint detection.

(Technically top-level variables, because `require` can pass along function scopes, but treating them as effectively global is fairly reasonable)

EDIT: This seems to work if `require 'src/b.php'` is found anywhere else, but it would still be useful to have a way for non-analyzable requires to be able to associate the files with each other.

```php
<?php
// src/b.php
echo $unsafe;
```

```php
<?php
// src/global.php
$unsafe = $_GET['unsafe'];
```

**Comment by TysonAndre** [comment]
Also, how does psalm infer that the variable declarations/union types in different files are related? What classes/methods are involved in doing that?

**Comment by muglug** [comment]
> Also, how does psalm infer that the variable declarations/union types in different files are related? What classes/methods are involved in doing that?

Assuming the files are linked using `include`/`require`, the `IncludeAnalyzer` should handle that.

> Would adding a custom plugin to the repo help in this use case if include `custom_abs_path('some_path.php')`; was the reason the files were related?

Absolutely, if that plugin set the return type of that function to be a literal string

**Comment by TysonAndre** [comment]
This can probably be closed if you don't think it's widely applicable - setting the return type in a plugin should be good enough and avoid excessive false positives for my use case.

**Comment by mmcev106** [comment]
[Click here](https://psalm.dev/r/90bd306d45) for a similar scenario involving globals within a single file.  If anyone has any ideas for how to support either scenario, I might be able to implement it.

**Comment by orklah** [comment]
Well, *technically*, all you would have to do is piggyback on the globals module that we already have to add custom globals. Now I'm not exactly sure about the order in which Psalm analyse statements in a global context so I'm not sure how it would play out

**Comment by mmcev106** [comment]
@orklah, I appreciate the quick feedback.  I've spent a few more hours today trying to understand the relevant parts of the psalm source, but am striking out I'm afraid...  I'm trying hard to develop some sort of smarter strategy for interacting with the source, but just feel like I always end up spinning my wheels and going in circles...  I'm all ears if anyone has any suggestions.
